lib/Parse/ParseExpr.cpp

```diff
@@ -2744,6 +2744,81 @@ void Parser::ParseBlockId(SourceLocation CaretLoc) {
   Actions.ActOnBlockArguments(CaretLoc, DeclaratorInfo, getCurScope());
 }
 
+ExprResult Parser::ParseBoundsExpression() {
+  if (Tok.getKind() != tok::identifier) {
+    // This can't be a contextual keyword that begins a bounds expression,
+    // so stop now.
+    Diag(Tok, diag::err_expected_bounds_expr);
+    return ExprError();
+  }
+
+  IdentifierInfo *Ident = Tok.getIdentifierInfo();
+  SourceLocation BoundsKWLoc = Tok.getLocation();
+  ConsumeToken();
+
+  // Parse the body of a bounds expression. Set Result to ExprError() if
+  // something goes wrong.
+  BalancedDelimiterTracker PT(*this, tok::l_paren);
+  if (PT.expectAndConsume(diag::err_expected_lparen_after, Ident->getNameStart()))
+    return ExprError();
+
+  ExprResult Result;
+  if (Ident == Ident_byte_count || Ident == Ident_count) {
+    // Parse byte_count(e1) or count(e1)
+    Result = Actions.CorrectDelayedTyposInExpr(ParseAssignmentExpression());
+    CountBoundsExpr::Kind CountKind = Ident == Ident_count ?
+      CountBoundsExpr::Kind::ElementCount : CountBoundsExpr::Kind::ByteCount;
+    if (!Result.isInvalid())
+      Result = Actions.ActOnCountBoundsExpr(BoundsKWLoc, CountKind,
+                                            Result.get(),
+                                           Tok.getLocation());
+    else Result = ExprError();
```

Comment:
See previous nit about else bodies (and ditto for other uses in this function).  Also feel free to tell me to stop bothering you about this sort of thing if this conforms to patterns in this code base...